pkg/p2p/server.go

```diff
@@ -0,0 +1,872 @@
+							// caught in tests, but in the same time we can provide better
+							// resilience in production (`DPanic` does not panic in production).
+							//
+							// Note: Even if `GracefulUnregister` does fail, the handle is still
+							// unregistered, only forcefully.
+							log.L().DPanic("failed to gracefully unregister handle",
+								zap.Error(err))
+						}
+						log.Debug("handler deregistered", zap.String("topic", task.topic))
+						if task.done != nil {
+							close(task.done)
+						}
+					}()
+				}
+			case taskOnRegisterPeer:
+				log.Debug("taskOnRegisterPeer",
+					zap.String("sender", task.sender.streamMeta.SenderId),
+					zap.Int64("epoch", task.sender.streamMeta.Epoch))
+				if err := m.registerPeer(ctx, task.sender, task.clientAddr); err != nil {
+					if cerror.ErrPeerMessageStaleConnection.Equal(err) || cerror.ErrPeerMessageDuplicateConnection.Equal(err) {
+						// These two errors should not affect other peers
+						if err1 := task.sender.Send(ctx, errorToRPCResponse(err)); err1 != nil {
+							return errors.Trace(err)
+						}
+						continue // to handling the next task
+					}
+					return errors.Trace(err)
+				}
+			case taskDebugDelay:
+				log.Info("taskDebugDelay started")
+				select {
+				case <-ctx.Done():
+					log.Info("taskDebugDelay canceled")
+					return errors.Trace(ctx.Err())
+				case <-task.doneCh:
+				}
+				log.Info("taskDebugDelay ended")
+			}
+		}
+	}
+}
+
+func (m *MessageServer) tick(ctx context.Context) {
+	var peersToDeregister []*cdcPeer
+	defer func() {
+		for _, peer := range peersToDeregister {
+			// err is nil because the peers are gone already, so sending errors will not succeed.
+			m.deregisterPeer(ctx, peer, nil)
+		}
+	}()
+
+	m.peerLock.RLock()
+	defer m.peerLock.RUnlock()
+
+	for _, peer := range m.peers {
+		var acks []*p2p.Ack
+		m.acksMapLock.RLock()
+		for topic, ack := range m.acksMap[peer.PeerID] {
+			acks = append(acks, &p2p.Ack{
+				Topic:   topic,
+				LastSeq: ack,
+			})
+		}
+		m.acksMapLock.RUnlock()
+		if len(acks) == 0 {
+			// No topic to ack, skip.
+			continue
+		}
+
+		peer.metricsAckCount.Inc()
+		err := peer.sender.Send(ctx, p2p.SendMessageResponse{
+			Ack:        acks,
+			ExitReason: p2p.ExitReason_OK,
+		})
+		if err != nil {
+			log.Warn("sending response to peer failed", zap.Error(err))
+			if cerror.ErrPeerMessageInternalSenderClosed.Equal(err) {
+				peersToDeregister = append(peersToDeregister, peer)
+			}
+		}
+	}
+}
+
+func (m *MessageServer) deregisterPeer(ctx context.Context, peer *cdcPeer, err error) {
+	log.Info("Deregistering peer",
+		zap.String("sender", peer.PeerID),
+		zap.Int64("epoch", peer.Epoch),
+		zap.Error(err))
+
+	m.peerLock.Lock()
+	delete(m.peers, peer.PeerID)
+	m.peerLock.Unlock()
+	if err != nil {
+		peer.abortWithError(ctx, err)
+	}
+}
+
+// MustAddHandler registers a handler for messages in a given topic and waits for the operation
+// to complete.
+func (m *MessageServer) MustAddHandler(
```

Comment:
```suggestion
func (m *MessageServer) AddHandler(
```

`Must` implies panic if AddHandler fails and should not return error.